plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufSchemaParser.java

```diff
@@ -0,0 +1,115 @@
+import io.trino.spi.type.TypeManager;
+
+import javax.inject.Inject;
+
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.TimestampType.createTimestampType;
+import static io.trino.spi.type.VarbinaryType.VARBINARY;
+import static io.trino.spi.type.VarcharType.createUnboundedVarcharType;
+import static java.util.Objects.requireNonNull;
+
+public class ProtobufSchemaParser
+        implements SchemaParser
+{
+    private static final String TIMESTAMP_TYPE_NAME = "google.protobuf.Timestamp";
+    private final TypeManager typeManager;
+
+    @Inject
+    public ProtobufSchemaParser(TypeManager typeManager)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+    }
+
+    @Override
+    public KafkaTopicFieldGroup parse(ConnectorSession session, String subject, ParsedSchema parsedSchema)
+    {
+        ProtobufSchema protobufSchema = (ProtobufSchema) parsedSchema;
+        return new KafkaTopicFieldGroup(
+                ProtobufRowDecoder.NAME,
+                Optional.empty(),
+                Optional.of(subject),
+                protobufSchema.toDescriptor().getFields().stream()
+                        .map(field -> new KafkaTopicFieldDescription(
+                                field.getName(),
+                                getType(field),
+                                field.getName(),
+                                null,
+                                null,
+                                null,
+                                false))
+                        .collect(toImmutableList()));
+    }
+
+    private Type getType(FieldDescriptor fieldDescriptor)
+    {
+        Type baseType = switch (fieldDescriptor.getJavaType()) {
+            case BOOLEAN -> BOOLEAN;
+            case INT -> INTEGER;
+            case LONG -> BIGINT;
+            case FLOAT -> REAL;
+            case DOUBLE -> DOUBLE;
+            case BYTE_STRING -> VARBINARY;
+            case STRING, ENUM -> createUnboundedVarcharType();
+            case MESSAGE -> getTypeForMessage(fieldDescriptor);
+        };
+
+        // Protobuf does not support adding repeated label for map type but schema registry incorrecty adds it
+        if (fieldDescriptor.isRepeated() && !fieldDescriptor.isMapField()) {
+            return new ArrayType(baseType);
+        }
+        return baseType;
+    }
+
+    private Type getTypeForMessage(FieldDescriptor fieldDescriptor)
+    {
+        Descriptor descriptor = fieldDescriptor.getMessageType();
+        if (fieldDescriptor.getMessageType().getFullName().equals(TIMESTAMP_TYPE_NAME)) {
+            return createTimestampType(6);
```

Comment:
Per https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Timestamp

> A Timestamp represents a point in time independent of any time zone or calendar, represented as seconds and fractions of seconds at nanosecond resolution in UTC Epoch time

I.e. it's an instant.
We don't have an instant type (https://github.com/trinodb/trino/issues/2273), so we represent point-in-time data as "timestamp with time zone".

Would you consider changing the mapping to reveal the point-in-time semantics of the data?

Reply:
Also, the type is nano-second precision. Why do we map it to microsecond precision?

cc @mx123

Reply:
These are direct ports from SEP, maybe @Praveen2112 remembers why?
Changing the semantic or precision might break back-compat, we might have to support both local/instant x all precisions combo?

Reply:
> Changing the semantic or precision might break back-compat,

Of course. Every change, including bug fixes, can break something.

>  we might have to support both local/instant x all precisions combo?

not sure why we would want that.